cache/cache_modifier.go

```diff
@@ -0,0 +1,232 @@
+	if !replay && hermetic {
+		return nil, fmt.Errorf("cache.Modifier: cannot use hermetic mode if not replaying")
+	}
+	if bucket == "" && (update || replay) {
+		return nil, fmt.Errorf("cache.Modifier: bucket name cannot be empty if updating or replaying")
+	}
+
+	opt := &bolt.Options{
+		Timeout: 10 * time.Second,
+	}
+	log.Debugf("cache.Modifier: opening boltdb file %q", filepath)
+	db, err := bolt.Open(filepath, 0600, opt)
+	if err != nil {
+		return nil, fmt.Errorf("cache.Modifier: bolt.Open(%q): %v", filepath, err)
+	}
+	// TODO: Figure out how to close the db after use.
+
+	if bucket != "" {
+		if err := db.Update(func(tx *bolt.Tx) error {
+			_, err := tx.CreateBucketIfNotExists([]byte(bucket))
+			if err != nil {
+				return fmt.Errorf("CreateBucketIfNotExists(%q): %s", bucket, err)
+			}
+			return nil
+		}); err != nil {
+			return nil, fmt.Errorf("cache.Modifier: db.Update(): %v", err)
+		}
+	}
+
+	return &modifier{
+		db:       db,
+		bucket:   bucket,
+		update:   update,
+		replay:   replay,
+		hermetic: hermetic,
+	}, nil
+}
+
+// modifierFromJSON parses JSON into cache.Modifier.
+//
+// Example JSON Configuration message:
+// {
+//   "file":     "/some/cache.db",
+//   "bucket":   "some_project",
+//   "update":   true,
+//   "replay":   true,
+//   "hermetic": false
+// }
+func modifierFromJSON(b []byte) (*parse.Result, error) {
+	msg := &modifierJSON{}
+	if err := json.Unmarshal(b, msg); err != nil {
+		return nil, fmt.Errorf("cache.Modifier: json.Unmarshal(): %v", err)
+	}
+
+	mod, err := NewModifier(msg.File, msg.Bucket, msg.Update, msg.Replay, msg.Hermetic)
+	if err != nil {
+		return nil, err
+	}
+	return parse.NewResult(mod, msg.Scope)
+}
+
+// Close closes the underlying database.
+func (m *modifier) Close() error {
+	if m.db != nil {
+		return m.db.Close()
+	}
+	return nil
+}
+
+// ModifyRequest modifies the http.Request.
+func (m *modifier) ModifyRequest(req *http.Request) error {
+	if !m.replay {
+		return nil
+	}
+
+	key, err := getCacheKey(req)
+	if err != nil {
+		return fmt.Errorf("cache.Modifier: getCacheKey(): %v", err)
+	}
+
+	if err := m.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(m.bucket))
+		cached := b.Get(key)
+		if cached != nil {
+			res, err := http.ReadResponse(bufio.NewReader(bytes.NewReader(cached)), req)
+			if err != nil {
+				return fmt.Errorf("http.ReadResponse(): %v", err)
+			}
+			ctx := martian.NewContext(req)
+			ctx.SkipRoundTrip()
+			ctx.Set(cachedResponseCtxKey, res)
+			return nil
+		} else if m.hermetic {
+			return fmt.Errorf("in hermetic mode and no cached response found")
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("cache.Modifier: %v", err)
+	}
+	return nil
+}
+
+// ModifyResponse modifies the http.Response.
+func (m *modifier) ModifyResponse(res *http.Response) error {
+	ctx := martian.NewContext(res.Request)
+	cached, ok := ctx.Get(cachedResponseCtxKey)
+	if ok {
+		*res = *cached.(*http.Response)
```

Comment:
should make res.Request the request that we're responding to - not the request associated with the cached response
rather than wholesale swapping out the response, might it be more clear to replace the body and the headers of `res`?

Reply:
Good catch about the res.Request.  Will update it.

For the response, I swapped it out entirely, since the one we start with is a canned response anyway.  Besides the body and headers, there's also the HTTP major/minor versions, and possible trailers.  Is there any concern with just swapping it out?  The only thing I can think of is if there is a response modifier upstream that had data it wanted to preserve.  However, those modifiers should be downstream from the cache modifier, so they can modify the cached response.